{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This script implements phase I in BMGI algorithm: more detials can be found in BMGI paper: \n",
    "\n",
    "Random Forest is used here to score every gene for classification power using its \\\n",
    "isoforms expression values as features.\n",
    "\n",
    "Required Inputs:\n",
    "1- Annotated gene sets. Each gene is represented by an Ensemble ID (e.g. ). \n",
    "2- TEM: sigle-cell processed transcriptomic expression data (rows are the samples (i.g. cells), columns are the transcripts). csv file\n",
    "3- GEM: sigle-cell processed gene expression data (rows are the samples (i.g. cells), columns are the genes). csv file\n",
    "4- Sample labels (samples (i.g. cells) belongs to classes, each samples should have a label).\n",
    "\n",
    "Not required inputs:\n",
    "5- P-value threshold (defult = 0.001).\n",
    "6- Percentage threshold (defult = 0.6). \n",
    "7- Number of trees for Random Forest (defult =100). \n",
    "8- Number of folds for cross validation (defult =10). \n",
    "\n",
    "Outputs:\n",
    "1- \n",
    "\n",
    "\n",
    "\n",
    "# Example:\n",
    "\n",
    "python Model/BMGI_Phase_I.py \\\n",
    "    --gene_sets    /home/maburid/AMIA_project/Transc_stem_cell.csv \\\n",
    "    --TEM_dataset  /home/maburid/AMIA_project/Transc_stem_cell.csv \\\n",
    "    --GEM_dataset  /home/maburid/AMIA_project/Gene_stem_cell.csv \\\n",
    "    --labels       /home/maburid/AMIA_project/labels.csv \\\n",
    "    --num_trees    50 \\\n",
    "    --cv           5\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import gc\n",
    "gc.collect()\n",
    "import numpy as np\n",
    "import sklearn\n",
    "import sklearn.ensemble\n",
    "import sklearn.mixture\n",
    "import scipy\n",
    "\n",
    "import random\n",
    "import sys\n",
    "import json\n",
    "import time\n",
    "import mygene\n",
    "import gseapy\n",
    "from gseapy import parser \n",
    "import pandas as pd                        \n",
    "import xlrd        \n",
    "import os\n",
    "from sklearn import preprocessing\n",
    "import operator\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # parse command-line arguments\n",
    "    parser = argparse.ArgumentParser(description=\"Select genes which perform significantly better than an equivalent random gene for each annotated set.\")\n",
    "    parser.add_argument(\"--gene_sets\", help=\"list of annotated gene sets\", required=True)\n",
    "    parser.add_argument(\"--TEM_dataset\", help=\"TEM input dataset (samples x transcripts)\", required=True)\n",
    "    parser.add_argument(\"--GEM_dataset\", help=\"GEM input dataset (samples x genes)\", required=True)\n",
    "    parser.add_argument(\"--labels\", help=\"list of sample labels\", required=True)\n",
    "    \n",
    "    parser.add_argument(\"--num_trees\", help=\"number of trees in random forest\", type=int, default=100)\n",
    "    parser.add_argument(\"--cv\", help=\"number of folds for k-fold cross validation\", type=int, default=10)\n",
    "    parser.add_argument(\"--p_threshold\", help=\"maximum p-value required for a gene to be selected\", type=float, default=0.001)\n",
    "    parser.add_argument(\"--percent_threshold\", help=\"minimum percentage of filtered genes in a set to the origional genes, otherwise, if less \\\n",
    "    than this threshould, dont count the set as a significant\", type=float, default=0.6)\n",
    "    \n",
    "    args = parser.parse_args()\n",
    "\n",
    "    # load input data\n",
    "    print(\"loading input datasets...\")\n",
    "    filename_TEM = args.TEM_dataset\n",
    "    filename_GEM = args.GEM_dataset \n",
    "    \n",
    "    Trans_data=pd.read_csv(filename_TEM)     \n",
    "    Gene_data=pd.read_csv(filename_GEM)        \n",
    "    print(\"Genes_level_data_shape:  \", Gene_data.shape )\n",
    "    print(\"Genes_level_data_shape:  \", Trans_data.shape )\n",
    "\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3_7_cpu",
   "language": "python",
   "name": "python_custom3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
